### ISSUE TITLE
MarkLogic representation and handling of "Undefined" values

### ISSUE THREAD

**Issue by wemrysi** [quasar-analytics/quasar#1502]
The `QScript` data model includes `Undefined` values which need a representation in MarkLogic as well as handling of the `MapFunc`s `Undefined` and `IfUndefined`.

Other functions may need modification to deal with undefined values as well once the semantics of undefined have been clarified.

**Comment by sellout** [comment]
Minor niggle, but technically, `Undefined` isn’t part of the data model. It shouldn’t need to be represented in MarkLogic, but the MapFuncs _do_ need to be handled.

**Comment by jdegoes** [comment]
`Undefined` is an error resulting from applying an operation to the wrong type. It's sole purpose in runtime values is to enable error recovery / exception handling via `IfUndefined`. Semantically, undefined values are invisible to operations except for `IfUndefined`.

**Comment by wemrysi** [comment]
@sellout @jdegoes Thanks for the clarification, though it would seem the presence of `IfUndefied` implies _some_ sort of representation, no?

**Comment by sellout** [comment]
@wemrysi Well, the thing is there’s no way to _store_ an undefined value. I.e., the representation of a map with an entry whose value is undefined is a map without that entry. Statically known things are handled by the compiler (deleting entries that are always undefined, propagating undefineds, etc.), but of course, you might find that in a particular record, you end up trying to `4 + "3"`, and then you end up with that result being undefined. So, the planner (at the MapFunc level) does need some way to say “ok, I’m holding onto a bad value”, but that is really the _lack_ of a value. In Spark this could be modeled by `Option[Data] => Option[Data]` (although currently isn’t). In most cases it would be simply handled with `(a1 ⊛ a2)(op(_, _)).join`, but Option can’t be the `M` in the fold, because in the cases of things like `MakeMap` and `IfUndefined`, you have to make an empty map, and return the second value instead.

tl;dr So, yeah, you need _some_ representation of an exceptional state while processing, but not anything that can end up in a data store.
